database/src/main/java/com/firebase/ui/database/IndexFirebaseArray.java

```diff
@@ -0,0 +1,138 @@
+        return mSnapshots.get(index);
+    }
+
+
+    private int getIndexForKey(String key) {
+        int index = 0;
+        for (DataSnapshot snapshot : mSnapshots) {
+            if (snapshot.getKey().equals(key)) {
+                return index;
+            } else {
+                index++;
+            }
+        }
+
+        throw new IllegalArgumentException("Key not found");
+    }
+
+    @Override
+    public void setOnChangedListener(OnChangedListener listener) {
+        mListener = listener;
+    }
+
+    @Override
+    protected void notifyChangedListeners(OnChangedListener.EventType type, final int index, int oldIndex) {
+        if (mListener != null) {
+            switch (type) {
+                case Added:
+                    final boolean[] isNewListener = {true};
+
+                    ValueEventListener valueEventListener = new ValueEventListener() {
+                        @Override
+                        public void onDataChange(DataSnapshot dataSnapshot) {
+                            if (dataSnapshot.getValue() != null) {
+                                if (isNewListener[0]) {
+                                    mSnapshots.add(index, dataSnapshot);
+                                    mListener.onChanged(OnChangedListener.EventType.Added, index, -1);
+
+                                    isNewListener[0] = false;
+                                } else {
+                                    mSnapshots.set(getIndexForKey(dataSnapshot.getKey()), dataSnapshot);
+                                    mListener.onChanged(OnChangedListener.EventType.Changed,
+                                                        getIndexForKey(dataSnapshot.getKey()),
+                                                        -1);
+                                }
+                            } else {
+                                Log.w("Firebase-UI", "Key not found at ref: " + dataSnapshot.getRef().toString());
+                            }
+                        }
+
+                        @Override
+                        public void onCancelled(DatabaseError databaseError) {
+                            notifyCancelledListeners(databaseError);
+                        }
+                    };
+
+                    DatabaseReference ref = mRef.child(internalGetItem(index).getKey());
+                    ref.addValueEventListener(valueEventListener);
+                    mValueEventListeners.put(ref, valueEventListener);
+                    break;
+                case Changed:
+                    // We don't care because the key's value shouldn't matter.
+                    break;
+                case Removed:
+                    DatabaseReference rmRef = mRef.child(mSnapshots.get(index).getKey());
+                    rmRef.removeEventListener(mValueEventListeners.get(rmRef));
+                    mValueEventListeners.remove(rmRef);
+                    mSnapshots.remove(index);
```

Comment:
No. The problem is still there. Because, `FirebaseArray.onChildRemoved()` really to call `mSnapshots.remove(index);`. After that, it call `Firebase.Array.notifyChangedListeners()` was override in `IndexFirebaseArray`.

To fix, override `onChildRemoved` on `IndexFirebaseArray`
```
    @Override
    public void onChildRemoved(DataSnapshot snapshot) {
        int index = getIndexForKey(snapshot.getKey());
        DatabaseReference rmRef = mRef.child(internalGetItem(index).getKey());
        rmRef.removeEventListener(mValueEventListeners.get(rmRef));
        mValueEventListeners.remove(rmRef);

        super.onChildRemoved(snapshot);
    }
```
and change following lines to:
```
                case Removed:
                    mListener.onChanged(OnChangedListener.EventType.Removed, index, -1);
                    break;
```